numpy/ma/core.py

```diff
@@ -3893,77 +3894,45 @@ def _delegate_binop(self, other):
                 return True
         return False
 
-    def __eq__(self, other):
-        """
-        Check whether other equals self elementwise.
-
-        """
+    def _comparison(self, other, op):
         if self is masked:
             return masked
         omask = getattr(other, '_mask', nomask)
-        if omask is nomask:
-            check = self.filled(0).__eq__(other)
-            try:
-                check = check.view(type(self))
-                check._mask = self._mask
-            except AttributeError:
-                # Dang, we have a bool instead of an array: return the bool
-                return check
+        odata = other if omask is nomask else filled(other, 0)
+        check = getattr(self.filled(0), op)(odata)
+
+        if isinstance(check, (bool_, bool)):
+            return masked if self._mask else check
+
+        if self._mask is nomask:
+            mask = omask
         else:
-            odata = filled(other, 0)
-            check = self.filled(0).__eq__(odata).view(type(self))
-            if self._mask is nomask:
-                check._mask = omask
-            else:
-                mask = mask_or(self._mask, omask)
-                if mask.dtype.names:
-                    if mask.size > 1:
-                        axis = 1
-                    else:
-                        axis = None
-                    try:
-                        mask = mask.view((bool_, len(self.dtype))).all(axis)
-                    except ValueError:
-                        mask = np.all([[f[n].all() for n in mask.dtype.names]
-                                       for f in mask], axis=axis)
-                check._mask = mask
+            mask = mask_or(self._mask, omask)
+            if mask.dtype.names:
+                # Structured array.  Mask element if all parts are masked.
+                mask = mask.view(bool_).reshape(mask.shape+(-1,)).all(-1)
```

Comment:
I know this is the old behavior, but just want to note that `any` makes more sense to me than `all`. 

With `all`, I think it causes the result of the comparison to depend on the data at a masked value, eg:
```python
>>> m = np.ma.mvoid((1,2), mask=(0,1), dtype='i4,f4')
>>> n = np.ma.mvoid((1,0), dtype='i4,f4')
>>> o = np.ma.mvoid((1,1), dtype='i4,f4')
>>> (n == m), (m ==n, o == m)
(True, masked, False)
```